scriptworker/context.py

```diff
@@ -131,99 +167,160 @@ def create_queue(self, credentials):
+    @property
+    def task_id(self):
+        """string: The running task's taskId."""
+        if self.claim_task:
+            return self.claim_task['status']['taskId']
 
-        """
-        log.debug(message.format(path=path))
-        makedirs(os.path.dirname(path))
-        with open(path, "w") as fh:
-            json.dump(contents, fh, indent=2, sort_keys=True)
+    @property
+    def run_id(self):
+        """string: The running task's runId."""
+        if self.claim_task:
+            return self.claim_task['runId']
 
     @property
-    def projects(self):
-        """dict: The current contents of ``projects.yml``, which defines CI configuration.
+    def claim_task(self):
+        """dict: The current or most recent claimTask definition json from the queue.
 
-        I'd love to auto-populate this; currently we need to set this from
-        the config's ``project_configuration_url``.
+        This contains the task definition, as well as other task-specific
+        info.
+
+        When setting ``claim_task``, we also set ``self.task`` and
+        ``self.credentials``, zero out ``self.reclaim_task`` and ``self.proc``.
 
         """
-        if self._projects:
-            return dict(deepcopy(self._projects))
+        return self._claim_task
 
-    @projects.setter
-    def projects(self, projects):
-        self._projects = projects
+    @claim_task.setter
+    def claim_task(self, claim_task):
+        self._claim_task = claim_task
+        self.reclaim_task = None
+        self.task_process = None
+        self.task = claim_task['task']
+        self.credentials = claim_task['credentials']
 
     @property
-    def event_loop(self):
-        """asyncio.BaseEventLoop: the running event loop.
+    def reclaim_task(self):
+        """dict: The most recent reclaimTask definition.
 
-        This fixture mainly exists to allow for overrides during unit tests.
+        This contains the newest expiration time and the newest temp credentials.
+
+        When setting ``reclaim_task``, we also set ``self.credentials``.
+
+        ``reclaim_task`` will be ``None`` if there hasn't been a claimed task yet,
+        or if a task has been claimed more recently than the most recent
+        reclaimTask call.
 
         """
-        if not self._event_loop:
-            self._event_loop = asyncio.get_event_loop()
-        return self._event_loop
+        return self._reclaim_task
 
-    @event_loop.setter
-    def event_loop(self, event_loop):
-        self._event_loop = event_loop
+    @reclaim_task.setter
+    def reclaim_task(self, value):
+        self._reclaim_task = value
+        if value is not None:
+            self.credentials = value['credentials']
 
-    async def populate_projects(self, force=False):
-        """Download the ``projects.yml`` file and populate ``self.projects``.
 
-        This only sets it once, unless ``force`` is set.
+def set_task_context_paths(task_context, task_num=None):
```

Comment:
It could. I don't see a huge difference between a function that takes a `TaskContext` object and acts on it, and a method that acts on `self` in a `TaskContext` object. We only call this once, and then the method is available to a bunch of functions that don't need it.

Reply:
> then the method is available to a bunch of functions that don't need it.

The method is available to a bunch of functions that don't need it, even when it's not attached to the `TaskContext` object.

What is your technique for deciding if a method belongs to an object or not? The method seems very closely related to `TaskContext` (it has `..._task_context_...` in the name), so perhaps that's an indication that it should be attached?

Reply:
If they don't import the function, it's not loaded. But I don't have strong feelings here, other than we have this function available across a ton of functions that don't need it, and we seem to be pushing towards smaller objects.

Reply:
> If they don't import the function, it's not loaded.

From a technical/performance level, this is probably true, but I'm not sure how much impact this has

> we have this function available across a ton of functions that don't need it, and we seem to be pushing towards smaller objects.

I'm :100: on pushing towards objects that have fewer responsibilities. However, associating/dissociating a function with an object doesn't decrease the scope of the object. 

However, it does increase friction of using the object, since the function is more verbose to write (why does it have `..._task_context_...` in the name if it's for task contexts?) and it makes it harder to tell what methods are available for an object, since dissociated functions could be hiding anywhere in a module.

------

I don't believe that leaving this dissociated will have a strong _impact_  the function of `scriptworker`, but I'm pretty confident that associating it is an unambiguous (though small) improvement to maintainability.